### ISSUE TITLE
Endless loop + OOM when cycle is present in cross references

### ISSUE THREAD

**Issue by andrewisplinghoff** [weaviate/weaviate#4527]
### How to reproduce this bug?

If you have a cycle within the cross-references of one of your collections and peform a nested query over them, the request fails and the Weaviate server seems to enter an endless loop until it goes OOM.

Here's some Python to reproduce the issue:
```python
auth_config = weaviate.auth.AuthApiKey(api_key=...)
client = weaviate.Client(
    url=...,
    auth_client_secret=auth_config
)

collection_name = 'ContainsCycle'
class_obj = {
    "class": collection_name,
    "properties": [
        {
            "name": "ref",
            "dataType": [collection_name]
        }
        ]
}
client.schema.delete_class(collection_name)
client.schema.create_class(class_obj)

uuids = []
with client.batch as batch:
    for i in range(2):
        uuids.append(batch.add_data_object(
                data_object={},
                class_name=collection_name
            ))
    
print(uuids)

client.data_object.reference.add(
        from_class_name=collection_name,
        from_uuid=uuids[0],
        from_property_name="ref",
        to_class_name=collection_name,
        to_uuid=uuids[1],
    )

client.data_object.reference.add(
        from_class_name=collection_name,
        from_uuid=uuids[1],
        from_property_name="ref",
        to_class_name=collection_name,
        to_uuid=uuids[0],
    )

response = (
    client.query
    .get(
        collection_name,
        [f'''ref {{ ... on {collection_name} {{
                ref {{ ... on {collection_name} {{
                    _additional {{
                    id
                    }}
                }}}}
        }}}}'''],
    )
    .with_limit(2).do()
)

print(json.dumps(response, indent=2))
```

### What is the expected behavior?

The GraphQL query should succeed with minimal delay.

### What is the actual behavior?

GraphQL query fails after 1 minute (I guess some timeout), server starts permanently using CPU (looks like an endless loop), uses more and more memory until it goes OOM.

### Supporting information

In the weaviate logs, the following error is printed:
{"action":"requests_total","api":"graphql","class_name":"","error":"context canceled","level":"error","msg":"unexpected error","query_type":"","time":"2024-03-21T15:57:44Z"}

The example python code fails with either a timeout error or the following:
  File "/home/cdsw/andre/venv/lib/python3.9/site-packages/weaviate/gql/filter.py", line 124, in do
    res = _decode_json_response_dict(response, "Query was not successful")
  File "/home/cdsw/andre/venv/lib/python3.9/site-packages/weaviate/util.py", line 798, in _decode_json_response_dict
    raise UnexpectedStatusCodeException(location, response)
weaviate.exceptions.UnexpectedStatusCodeException: Query was not successful! Unexpected status code: 504, with response body: None.

The code succeeds when called with `with_limit(1)`, so the problem only happens when an entity is retrieved that was already part of the previous result set (retrieved via the cross reference).

I'm using Weaviate client 3.26.2, but that should not be relevant, it also happens when sending the GraphQL query using GraphiQL.

### Server Version

1.24.4

### Code of Conduct

- [X] I have read and agree to the Weaviate's [Contributor Guide](https://weaviate.io/developers/contributor-guide) and [Code of Conduct](https://weaviate.io/service/code-of-conduct)

**Comment by andrewisplinghoff** [comment]
With this minimal example, the server did not actually end up Out of Memory for us, but we saw it going OOM earlier due to this bug when using collections with a bit more data in them. With this minimal example, it starts using 700-1100MB of memory after the query was tried and stays at that level in our environment (OpenShift).

**Comment by etiennedi** [comment]
Thanks for the report! We'll look into it.

**Comment by andrewisplinghoff** [comment]
Fixed in 1.23.14 & 1.24.6, I verified that the bug is resolved in 1.24.6. Thanks for the quick bugfix!
